src/core/ensemble_scheduler.cc

```diff
@@ -143,6 +143,20 @@ class EnsembleContext {
   std::shared_ptr<InferRequestProvider> request_provider_;
   std::shared_ptr<InferResponseProvider> response_provider_;
   std::function<void(Status)> OnComplete_;
+
+  // Each context should have its own label provider that is a superset of
+  // the ensemble servable's label provider. This label provider will provide
+  // labels from inner model if the labels for an output is not specified in
+  // the ensemble's label provider.
+  // Note that the context doesn't use the ensemble's label provider directly,
+  // because each context is a "snapshot" of the models when a request arrives.
+  // Different contexts of the same ensemble may have different labels due to
+  // model unload/reload, and this won't be reflected if they use the
+  // same label provider.
+  LabelProvider context_label_provider_;
+
+  // Output tensors whose labels are not provided by the ensemble servable
+  std::unordered_map<size_t, std::string> no_label_tensors_;
```

Comment:
I think you  need a map from each ensemble tensor to the LabelProvider associated with that tensor, if any.